Revit_Core_Engine/Query/UniqueParameterNames.cs

```diff
@@ -0,0 +1,179 @@
+        /***************************************************/
+
+        [Description("Returns unique parameter names for the collection of the elements.")]
+        [Input("elements", "Elements to get the parameter names from.")]
+        [Output("paramterNames", "Output containing instance parameter names as Item1 and type paramter names as Item2.")]
+        public static Output<HashSet<string>, HashSet<string>> UniqueParametersNames(this IEnumerable<Element> elements)
+        {
+            Output<HashSet<string>, HashSet<string>> parameterNames = new Output<HashSet<string>, HashSet<string>>
+            {
+                Item1 = new HashSet<string>(),
+                Item2 = new HashSet<string>()
+            };
+
+            if (elements == null || !elements.Any())
+                return parameterNames;
+
+            Dictionary<Document, List<Element>> elementsByDocumentDictionary = elements.ElementsByDocumentDictionary();
+
+            foreach (var elementsInDocPair in elementsByDocumentDictionary)
+            {
+                Document doc = elementsInDocPair.Key;
+                List<Element> elementsFromOneDocument = elementsInDocPair.Value;
+
+                if (!elementsFromOneDocument.Any())
+                    continue;
+
+                Dictionary<ElementId, List<Element>> elementsByCategoryDictionary = elementsFromOneDocument.ElementsByCategoryDictionary();
+
+                foreach (var elementsByCat in elementsByCategoryDictionary)
+                {
+                    List<Element> elementsOfCat = elementsByCat.Value;
+
+                    if (!elementsOfCat.Any()) 
+                        continue;
+
+                    HashSet<int> instanceParameterIds = elementsOfCat.UniqueParametersIds();
+                    IEnumerable<string> instanceParameterNames = instanceParameterIds.Select(x => doc.ParameterName(x));
+
+                    foreach (string name in instanceParameterNames)
+                        parameterNames.Item1.Add(name);
+
+                    IEnumerable<Element> elementTypes = elementsOfCat.UniqueTypeIds().Select(x => doc.GetElement(new ElementId(x)));
+                    HashSet<int> typeParametereIds = elementTypes.UniqueParametersIds();
+                    IEnumerable<string> typeParameterNames = typeParametereIds.Select(x => doc.ParameterName(x));
+
+                    foreach (string name in typeParameterNames)
+                        parameterNames.Item2.Add(name);
+                }
+            }
+
+            return parameterNames;
+        }
+
+        /***************************************************/
+
+        [Description("Creates dictionary for elements per document.")]
+        private static Dictionary<Document, List<Element>> ElementsByDocumentDictionary(this IEnumerable<Element> elements)
+        {
+            Dictionary<Document, List<Element>> elementsByDoc = new Dictionary<Document, List<Element>>();
```

Comment:
Probably no effect on performance, but you would encode it in a single line by chaining Linq `GroupBy` and `ToDictionary`